### ISSUE TITLE
gitstatus_stop missing for zsh plugin

### ISSUE THREAD

**Issue by bew** [romkatv/gitstatus#27]
Also I've noticed that `gitstatus_check` only checks if a previous `gitstatus_start` succeeded but it doesn't actually check if the daemon is currently running.

Would you accept a PR to update both? (close & check)

**Comment by romkatv** [comment]
Could you describe what you need `gitstatus_stop` for? The only use-case for it that I know is to be able to restart gitstatusd when you source `~/.zshrc` for the second time. What's yours?

As far as `gitstatus_check`, I'd like to remove it and instead document error codes in `gitstatus_query`. Why do you want to use it?

**Comment by romkatv** [comment]
FYI: I added `gitstatus_stop` to ZSH bindings.

My question about `gitstatus_check` still stands. Could you describe why the existing behavior doesn't work for you and what you want to achieve with the proposed change?

**Comment by bew** [comment]
Thanks! Sorry for the delay.. 
Basically I'd like to stop the daemon when I run some specific commands, or after some time when I'm not browsing git repos for example. And have a way to check if it's currently running or not would help I think.
Of course I can check myself with the pid, but it'd be easier/better encapsulated if it was done in the check "method".

Does that answer your question?

**Comment by romkatv** [comment]
> Basically I'd like to stop the daemon when I run some specific commands, or after some time when I'm not browsing git repos for example.

When you aren't calling `gitstatus_query`, gitstatus isn't using any resources. It'll have zero CPU load and won't touch its memory, which will allow the OS to swap it out. There is little need to stop it.

If you still want to stop gitstatus, the best way to do it is by calling `gitstatus_stop`. A more complex and less robust way (not recommended) is `gitstatus_check && gitstatus_stop`. Even more complex and even less robust (absolutely not recommended) is `[[ -n $GITSTATUS_DAEMON_PID_MY_ID ]] && kill -0 $GITSTATUS_DAEMON_PID_MY_ID && gitstatus_stop`.

Likewise, the best way to start gitstatus is `gitstatus_start`. Calling `gitstatus_check` or poking `$GITSTATUS_DAEMON_PID_MY_ID` will make code more complex and less robust.

> Of course I can check myself with the pid, but it'd be easier/better encapsulated if it was done in the check "method".

`gitstatus_start` and `gitstatus_stop` bring up/down more than just a process. There are locks, pipes, file descriptors, etc. All of these resources are encapsulated and the user doesn't need to bother with the details. All resources are brought up/down together as if in a single transaction. `gitstatus_check` tells you whether the whole bunch of resources is up/down.

If `gitstatus_check` ever disagrees with `[[ -n $GITSTATUS_DAEMON_PID_MY_ID ]] && kill -0 $GITSTATUS_DAEMON_PID_MY_ID`, this indicates a bug. It cannot be worked around by the user code. The equivalence of these two commands is one of the many invariants that the code maintains. However, these invariants are undocumented, for the aren't a part of the public API. There is no guarantee they will hold in the future. What is guaranteed is the behavior of `gitstatus_start`, `gitstatus_stop` and `gitstatus_query` (and `gitstatus_check`; although, as I mentioned, this function is almost useless and I'd like to remove it).

To summarize, you probably don't need to stop gitstatus. If you really do need to stop gitstatus, use `gitstatus_stop` to do it, and use `gitstatus_start` to start. If you find yourself in a situation where calling `gitstatus_check` or poking `$GITSTATUS_DAEMON_PID_MY_ID` actually helps with achieving any goal, please open an issue.

Does this make sense?

**Comment by romkatv** [comment]
Closing, as I've added `gitstatus_stop` and `gitstatus_check` is working as intended. If you encounter a use case that cannot be implemented (or is difficult to implement) with the current gitstatus API, but _can be_ (easier) implemented if the API is changed and/or extended, please either reopen this issue or open another.

**Comment by bew** [comment]
Yes it makes sense, I'll need some time to digest this, ans see how I'm going to work with it. Thanks for the explanation though!
